dbms/src/Common/BackgroundSchedulePool.cpp

```diff
@@ -0,0 +1,259 @@
+}
+
+
+BackgroundSchedulePool::~BackgroundSchedulePool()
+{
+    try
+    {
+        shutdown = true;
+        wakeup_event.notify_all();
+        queue.wakeUpAll();
+
+        delayed_thread.join();
+
+        LOG_TRACE(&Logger::get("BackgroundSchedulePool"), "Waiting for threads to finish.");
+        for (std::thread & thread : threads)
+            thread.join();
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+}
+
+
+BackgroundSchedulePool::TaskHandle BackgroundSchedulePool::addTask(const std::string & name, const Task & task)
+{
+    return std::make_shared<TaskInfo>(*this, name, task);
+}
+
+
+void BackgroundSchedulePool::removeTask(const TaskHandle & task)
+{
+    task->deactivate();
+}
+
+
+void BackgroundSchedulePool::scheduleDelayedTask(const TaskHandle & task, size_t ms, std::lock_guard<std::mutex>&)
+{
+    Poco::Timestamp current_time;
+
+    {
+        std::lock_guard lock(delayed_tasks_lock);
+
+        if (task->delayed)
+            delayed_tasks.erase(task->iterator);
+
+        task->iterator = delayed_tasks.emplace(current_time + (ms * 1000), task);
+        task->delayed = true;
+    }
+
+    wakeup_event.notify_all();
+}
+
+
+void BackgroundSchedulePool::cancelDelayedTask(const TaskHandle & task, std::lock_guard<std::mutex> &)
+{
+    {
+        std::lock_guard lock(delayed_tasks_lock);
+        delayed_tasks.erase(task->iterator);
+        task->delayed = false;
+    }
+
+    wakeup_event.notify_all();
+}
+
+
+void BackgroundSchedulePool::threadFunction()
+{
+    setThreadName("BackgrSchedPool");
+
+    MemoryTracker memory_tracker;
+    memory_tracker.setMetric(CurrentMetrics::MemoryTrackingInBackgroundSchedulePool);
+    current_memory_tracker = &memory_tracker;
+
+    while (!shutdown)
+    {
+        if (Poco::AutoPtr<Poco::Notification> notification = queue.waitDequeueNotification())
+        {
+            TaskNotification & task_notification = static_cast<TaskNotification &>(*notification);
+            task_notification.execute();
+        }
+    }
+
+    current_memory_tracker = nullptr;
+}
+
+
+void BackgroundSchedulePool::delayExecutionThreadFunction()
+{
+    setThreadName("BckSchPoolDelay");
+
+    while (!shutdown)
+    {
+        TaskHandle task;
+
+        {
+            std::unique_lock lock(delayed_tasks_lock);
+            Poco::Timestamp min_time;
+
+            if (!delayed_tasks.empty())
+            {
+                auto t = delayed_tasks.begin();
+                min_time = t->first;
+                task = t->second;
+            }
+
+            if (!task)
+            {
+                wakeup_event.wait(lock);
+                continue;
+            }
+
+            Poco::Timestamp current_time;
+
+            if (min_time > current_time)
+            {
+                wakeup_event.wait_for(lock, std::chrono::microseconds(min_time - current_time));
+                continue;
```

Comment:
This pattern is a bit strange - after the thread is awakened, it will be granted delayed_tasks_lock, then it will release it, only to try and immediately reacquire it on line 230.

But probably doesn't matter for correctness.

Reply:
You are right that in this case the lock is released and then reacquired. But in this case we cannot just schedule the task. Maybe was removed from the queue so for this reason we are going again through checks. 

I need to think if calling active() method on the task is enough or there are also other implications.

Reply:
Yes we need to go through all the checks after returning from the wait() method (even if for the possibility of a spurious wakeup).

My point was that it is not necessary to release the lock and immediately re-acquire it to check the condition - the lock is already acquired after we return from the wait method().

Reply:
@ztlpn I'm afraid is not clear for me on how you want this fixed : I mean how I can  go through all the checks while also I'm not releasing the lock which is already acquired. 

Can you provide some hints ?

Reply:
Possible solution:
Here is the current code (in pseudocode)
```c++
while (!shutdown)
{
    {
        acquire lock;
        
        if (shutdown)
            break;

        if (no tasks)
        {
            condition.wait();
            continue;
        }
    }

    schedule task;
}
```
Every time we do `continue;` we release the lock only to immediately re-acquire it.

Now what if we add a second loop:
```c++
while (!shutdown)
{
    {
        acquire lock;

        while (!shutdown)
        {
            if (no tasks)
            {
                condition.wait();
                continue;
            }
        }
    }

    schedule task;
}
```
This way we still hold the lock after continuing and we execute the inner loop until we get a task to schedule. The lock is released only to wait on the condition.

What do you think?

Reply:
Oh I got your point now. I'll fix this tomorrow

Reply:
BTW there is a bug in the exit predicate (line 265) - should be `if (shutdown)` not `if(!shutdown)`...

Reply:
fixed in 7d6268c8a9d2074ed317b0ade21da210347c2af4

Reply:
You see any issue with this implementation :

```c++
void BackgroundSchedulePool::delayExecutionThreadFunction()
{
    setThreadName("BckSchPoolDelay");

    while (!shutdown)
    {
        TaskHandle task;
        bool found = false;

        {
            std::unique_lock lock(delayed_tasks_lock);

            while(!shutdown)
            {
                Poco::Timestamp min_time;

                if (!delayed_tasks.empty())
                {
                    auto t = delayed_tasks.begin();
                    min_time = t->first;
                    task = t->second;
                }

                if (!task)
                {
                    wakeup_cond.wait(lock);
                    continue;
                }

                Poco::Timestamp current_time;

                if (min_time > current_time)
                {
                    wakeup_cond.wait_for(lock, std::chrono::microseconds(min_time - current_time));
                    continue;
                }
                else
                {
                    //we have a task ready for execution
                    found = true;
                    break;
                }
            }
        }

        if(found)
            task->schedule();
    }
}
```

Reply:
Looks good.

Reply:
Fixed in 0a057695aae1a5764723294195f26aa995e9747b